### ISSUE TITLE
[Toothpick] Disconnect does not work

### ISSUE THREAD

**Issue by paulmiddelkoop** [raycast/extensions#3636]
### Extension

https://www.raycast.com/briankosw/toothpick

### Description

Disconnecting does not work most of the times. The extension says "Device disconnected successfully.", but the device is still connected.

This is a common problem across other Bluetooth utilities also. See [blueutil](https://github.com/toy/blueutil/blob/master/blueutil.m#L1044) and [BluetoothConnector](https://github.com/lapfelix/BluetoothConnector/issues/20) issues.

The problem is that `IOBluetoothDevice.closeConnection` returns successfully but does not close anything. 
In blueutil they have fixed it like [this](https://github.com/toy/blueutil/blob/master/blueutil.m#L1044). 

When I use `--disconnect` combined with `--wait-disconnect` it consistently works to disconnect.

```
blueutil --disconnect 04-fe-a1-e4-73-75 --wait-disconnect 04-fe-a1-e4-73-75
```
I hope this information helps with fixing the problem.

**Comment by VladCuciureanu** [comment]
_Notice: If anyone with more Obj-C experience sees this and finds any discrepancies, please do tell._
\<rant\> I like Apple. I love my M1 Max beast. With that said... Their _public/documented_ bluetooth API sucks. \<\/rant\>

So, I looked into it 😮‍💨

What I think happens here is that there's a bug on Apple's side which gets accidentally fixed when subscribing to the device's disconnect event. _blueutil_ just treats the device as an 'IOBluetoothDevice', it subscribes to the event and waits for a certain timeout period after which it throws an error if no response arrived in the specified period of time. Nothing fancy (as far as I know). Even the blueutil's author was surprised this was a fix.

That said, there's nothing I can do right now. This is due to the fact that Raycast's AppleScript can't understand/access IOBluetoothDevices without the app itself (Raycast) declaring the need for bluetooth perms inside the App Sandbox. I have an implementation draft ready for testing but for now the fate of this lies in Raycast Team's hands.

tldr: Needs mods on Raycast's side. Messaged the dev team - now we wait.

**Comment by sjdonado** [comment]
> _Notice: If anyone with more Obj-C experience sees this and finds any discrepancies, please do tell._ <rant> I like Apple. I love my M1 Max beast. With that said... Their _public/documented_ bluetooth API sucks. </rant>
> 
> So, I looked into it 😮‍💨
> 
> What I think happens here is that there's a bug on Apple's side which gets accidentally fixed when subscribing to the device's disconnect event. _blueutil_ just treats the device as an 'IOBluetoothDevice', it subscribes to the event and waits for a certain timeout period after which it throws an error if no response arrived in the specified period of time. Nothing fancy (as far as I know). Even the blueutil's author was surprised this was a fix.
> 
> That said, there's nothing I can do right now. This is due to the fact that Raycast's AppleScript can't understand/access IOBluetoothDevices without the app itself (Raycast) declaring the need for bluetooth perms inside the App Sandbox. I have an implementation draft ready for testing but for now the fate of this lies in Raycast Team's hands.
> 
> tldr: Needs mods on Raycast's side. Messaged the dev team - now we wait.

There is a workaround that worked for me:
1. Look for this panel in System Settings: `Allow applications to access Bluetooth`
![image](https://user-images.githubusercontent.com/27580836/216760517-42d3f9b0-cda8-4760-800c-cb4c6cbd5575.png)
2. Click on the `+` button and select Raycast
![image](https://user-images.githubusercontent.com/27580836/216760557-13b318c0-c890-4f21-ad3d-5645a6059d5a.png)
3. Click on Quick & Re-open

**Comment by VladCuciureanu** [comment]
That's an amazing find! I'll redirect people to your post until a better solution comes up! Thanks Juan for sharing it!

**Comment by VladCuciureanu** [comment]
So, good news (kinda).
#8193
I implemented a compat-layer for 'blueutil' which improves this situation.

1. Install 'blueutil' via brew
2. Update Raycast
3. Update Toothpick
4. Open Toothpick preferences
5. Switch Bluetooth Backend to 'blueutil'

Connecting/disconnecting should be faster and less flaky. Testing was limited so I can't guarantee that it's perfect but, to me, it looks like an improvement :>
